src/mysql.c

```diff
@@ -238,6 +240,13 @@ static int mysql_config (oconfig_item_t *ci) /* {{{ */
 
 static MYSQL *getconnection (mysql_database_t *db)
 {
+	// This is necessary because we might be running on a different
+	// thread than before.
+	static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
+	pthread_mutex_lock(&mutex);
+	mysql_library_init(0, NULL, NULL);
```

Comment:
From the source code package "mysql-connector-c-6.1.6-src":

Evidence:
* include/mysql.h
```
#define mysql_library_init mysql_server_init
```
* libmysql/libmysql.c
```
    109 int STDCALL mysql_server_init(int argc __attribute__((unused)),
    110                               char **argv __attribute__((unused)),
    111                               char **groups __attribute__((unused)))
    112 {
    113   int result= 0;
    114   if (!mysql_client_init)
    115   {
...
    168   else
    169     result= (int)my_thread_init();         /* Init if new thread */
```
* mysys/my_thr_init.c
```
    257 my_bool my_thread_init()
    258 {
...
    280   mysql_mutex_lock(&THR_LOCK_threads);
    281   tmp->id= ++thread_id;
    282   ++THR_thread_count;
    283   mysql_mutex_unlock(&THR_LOCK_threads);
    284   set_mysys_thread_var(tmp);
```
* then in same file...
```
    354 int set_mysys_thread_var(struct st_my_thread_var *mysys_var)
    355 {
    356   if (THR_KEY_mysys_initialized)
    357     return my_set_thread_local(THR_KEY_mysys, mysys_var);
    358   return 0;
    359 }
```
* include/my_thread_local.h
```
     56 static inline int my_set_thread_local(thread_local_key_t key,
     57                                       void *value)
     58 {
     59 #ifdef _WIN32
     60   return !TlsSetValue(key, value);
     61 #else
     62   return pthread_setspecific(key, value);
     63 #endif
     64 }
```
Ergo, the thread local storage is not initialized for this thread unless you call this init routine on the thread. What goes wrong if the thread local storage is not set up? Well the first thing that goes wrong is...

* mysys/charset.c 
```
    487 static void init_available_charsets(void)
    488 {
...
    511   my_read_charset_file(&loader, fname, MYF(0));
```
* same file
```
    413 my_read_charset_file(MY_CHARSET_LOADER *loader,
...
    422   if (!my_stat(filename, &stat_info, MYF(myflags)) ||
    423        ((len= (uint)stat_info.st_size) > MY_MAX_ALLOWED_BUF) ||
    424        !(buf= (uchar*) my_malloc(key_memory_charset_file,
    425                                  len,myflags)))
```
* mysys/my_lib.c 
```
    359 MY_STAT *my_stat(const char *path, MY_STAT *stat_area, myf my_flags)
    360 {
...
    378   my_errno= errno;
```
* And what's so bad about that line? Well... see include/my_thread_local.h 
```
    102 #define my_errno mysys_thread_var()->thr_errno
```
* Which takes us back to mysys/my_thr_init.c
```
    346 struct st_my_thread_var *mysys_thread_var()
    347 {
    348   if (THR_KEY_mysys_initialized)
    349     return  (struct st_my_thread_var*)my_get_thread_local(THR_KEY_mysys);
    350   return NULL;
    351 }
```
* include/my_thread_local.h
```
     47 static inline void* my_get_thread_local(thread_local_key_t key)
     48 {
     49 #ifdef _WIN32
     50   return TlsGetValue(key);
     51 #else
     52   return pthread_getspecific(key);
     53 #endif
     54 }
```
So if the thread local variable hasn't been initialized *for this thread*, this returns a NULL, which segfaults at mysys/my_lib.c:378

Doing one init function before the threads are spawned, as suggested by the documentation, accomplishes nothing towards getting these thread local storage items initialized.

Does it still feel ugly? I am open to suggestions.